src/Raven.Server/Documents/Handlers/ShardedDocumentHandler.cs

```diff
@@ -25,29 +35,128 @@ public async Task Get()
+                        ContextPool.AllocateOperationContext(out TransactionOperationContext ctx);
+                        CmdsContext.Add(ctx);
 
-                var cmd = new ShardedCommand
+                        var task = ShardedContext.RequestExecutors[shard.Key].ExecuteAsync(cmd, ctx);
+                        Task.Add(task);
+                        Cmds.Add(shard.Key, cmd);
+                    }
+
+                    await System.Threading.Tasks.Task.WhenAll(Task);
+
+                    HttpContext.Response.StatusCode = (int)ShardedStatusCodeChooser.GetStatusCode(Cmds.Values);
+
+                    if (HttpContext.Response.IsSuccessStatusCode() == false)
+                        return;
+
+                    var actualEtag = ComputeHttpEtags.CombineEtags(EnumerateEtags(Cmds));
+
+                    if (etag == actualEtag)
+                    {
+                        HttpContext.Response.StatusCode = (int)HttpStatusCode.NotModified;
+                        return;
+                    }
+
+                    HttpContext.Response.Headers[Constants.Headers.Etag] = "\"" + actualEtag + "\"";
+
+                    using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBodyStream(), ServerStore.ServerShutdown))
+                    {
+                        writer.WriteStartObject();
+
+                        WriteArray(writer, nameof(GetDocumentsResult.Results), ids, shards);
+
+                        //TODO: write includes
+
+                        writer.WriteEndObject();
+                        await writer.OuterFlushAsync();
+                        //TODO: Add performance hints
+                    }
+                }
+            }
+            finally
+            {
+                Task.Clear();
+                Cmds.Clear();
+
+                Parallel.ForEach(CmdsContext, c =>
                 {
-                    Method = HttpMethod.Get,
-                    Url = $"/docs?id={Uri.EscapeUriString(id)}",
-                };
-                await ShardedContext.RequestExecutors[index].ExecuteAsync(cmd, context);
-                string responseEtag = cmd.Response?.Headers?.ETag?.Tag;
-                if (responseEtag != null)
-                    HttpContext.Response.Headers.Add(Constants.Headers.Etag, responseEtag);
+                    c.Dispose();
+                });
+            }
+        }
 
-                HttpContext.Response.StatusCode = (int)cmd.StatusCode;
-                //TODO: Pass the ETag
-                cmd.Result?.WriteJsonTo(ResponseBodyStream());
+        private IEnumerable<string> EnumerateEtags(Dictionary<int, ShardedCommand> cmds)
+        {
+            foreach (var cmd in cmds.Values)
+            {
+                yield return cmd.Response?.Headers?.ETag?.Tag;
+            }
+        }
+
+        private int WriteArray(AsyncBlittableJsonTextWriter writer, string arrayName, IEnumerable<string> ids,
```

Comment:
We should be able to simplify this

Reply:
If we need to keep the load order how can we simplify this?
@ppekrol do we need to keep load order in LoadOperation results?
Breaking down commands and aggregate their results as blittables is going to be complex for sure and error-prone.

Reply:
Got some ideas on this, coming over